words/core.asm

```diff
@@ -2966,68 +2941,66 @@ z_less_than:    rts
         ; Compile-only word to store TOS so that it is pushed on stack
         ; during runtime. This is a immediate, compile-only word. At runtime,
-        ; it works by calling literal_runtime by compling JSR LITERAL_RT.
+        ; it works by calling literal_runtime or compiling an inline equivalent.
         ;
         ; Note the cmpl_ routines use TMPTOS
         ; """
 xt_literal:
                 jsr underflow_1
 w_literal:
-                lda #template_push_tos_size
-                jsr check_nc_limit
-                bcc _inline
+                jsr literal_runtime
+                .word literal_runtime
+
+                ; ( n call-addr )
+                lda 3,x                         ; is it a byte value?
+                bne +
+
+                jsr two_literal_runtime
+                .word template_push_tos_size-2  ; TOS
+                .word template_push_tos+2       ; NOS, if we're inlining
+                bra _cmpl
++
+                jsr two_literal_runtime
+                .word template_push_tos_size    ; TOS
+                .word template_push_tos         ; NOS, if we're inlining
+_cmpl:
+                ; ( n call-addr inline-addr inline-sz )
 
-                ldy #>literal_runtime
-                lda #<literal_runtime
-                jsr cmpl_subroutine
+                jsr cmpl_by_limit2
+                ; ( n )
+                bcc _inline                     ; C=0 if inlined
 
-                ; Compile the value that is to be pushed on the Stack during
-                ; runtime
+                ; Compile the value to be pushed to data stack at runtime
                 jsr w_comma
-                bra z_literal
+_done:
+                sec                             ; tell w_two_literal we didn't inline
+                rts
 
 _inline:
-                ; we'll need the MSB (if non-zero) and LSB to fill in the template
-                ; which we set up on the stack in reverse order
-                ; first we need the STZ/STY opcode for the end of the template
-                ldy #$94        ; STY opcode
-                lda 1,x         ; MSB
-                bne +
-                ldy #$74        ; STZ opcode
-+               phy
-
-                lda 0,x         ; LSB
-                pha
-
-                ; if MSB is non-zero, stack it, otherwise skip first two bytes of template
-                ldy #2
-                lda 1,x         ; MSB
-                beq _copy
-                ldy #0
-                pha
+                ; update the placeholders in the template we compiled
+                ; temporarily reduce cp by 256 to make reverse indexing easier
+                dec cp+1
+                ; update the template
+                lda 1,x                         ; MSB non-zero?
+                beq +
+                ldy #256-9
+                sta (cp),y                      ; ldy #<MSB>
+                lda #$94                        ; opcode for STY
```

Comment:
you mean this post-compilation fixup, where we switch STZ to STY?

we would still need to write the actual byte/word values into template.  the old version had a custom template copying loop that filled placeholders from the stack, but this seemed shorter.

but I agree that this still doesn't seem ideal, I will have a think about it.